cosmicfishpie/LSSsurvey/photo_obs.py

```diff
@@ -686,6 +692,83 @@ def computecls(self):
 
         return cls
 
+    def computecls_vectorized(self):
+        """Vectorized function to compute the angular power spectrum for all observables, redshift bins and multipoles.
+
+        Returns
+        -------
+        dict
+            a dictionary containing all auto and cross correlation angular power spectra. Its keys are formatted to have an array of the power spectra in 'X ixY j'. Also the multipoles for which the angular power spectra were computed for are found in 'ells'.
+
+        Note
+        -----
+        Implements the following equation:
+
+            .. math::
+                C_{i,j}^{X,Y}(\\ell) = c \\int \\mathrm{d}z \\frac{W_i^X (z)W_j^Y (z)}{ H(z) r^2(z)}
+                P_{\\delta \\delta} \\big[\\frac{\\ell + 1/2}{r(z)} , z \\big]
+        """
+        upt.time_print(
+            feedback_level=self.feed_lvl,
+            min_level=2,
+            text="Computing Cls integral for {}".format(self.observables),
+            instance=self,
+        )
+
+        full_ell = self.ell
+        cls = {"ells": full_ell}
+        hub = self.cosmo.Hubble(self.z)
+
+        tstart = time()
+
+        # Pre-compute all clsintegrals
+        clinterps = {}
+        for obs1, obs2, bin1, bin2 in product(
+            self.observables, self.observables, self.binrange, self.binrange
+        ):
+            clinterps[(obs1, obs2, bin1, bin2)] = self.clsintegral(obs1, obs2, bin1, bin2, hub)
+
+        # Vectorized computation of finalcls
+        for (obs1, obs2, bin1, bin2), clinterp in clinterps.items():
+            lmin1, lmax1 = cfg.specs[f"lmin_{obs1}"], cfg.specs[f"lmax_{obs1}"]
+            lmin2, lmax2 = cfg.specs[f"lmin_{obs2}"], cfg.specs[f"lmax_{obs2}"]
+
+            mask = (
+                (full_ell >= lmin1)
+                & (full_ell <= lmax1)
+                & (full_ell >= lmin2)
+                & (full_ell <= lmax2)
+            )
+
+            finalcls = np.zeros(len(full_ell))
+            finalcls[mask] = clinterp(full_ell[mask])
```

Comment:
if we have the mask here do we need to keep it in the clsintegral?

Reply:
let's review that later!